### DISCUSSION TITLE
Biomeidfixer produces json file with incorrect assignments

### DISCUSSION THREAD

**Discussion by EricLWeiss** [desagas/Biome-Id-Fixer#15] [Q&A]
Biomeidfixer is scrambling the biomes of a world as soon as the mod is added, producing a json file with wildly incorrect biome assignments.  When we remove the mod and the json file the world's biomes go back to normal.  

Oddly, when we put biomeidfixer back into the mods folder it scrambles the biomes in precisely the same way, and produces the same (wrongly assigned) json file.

Why is this, and more importantly how can we get biomeidfixer to generate a json file with accurate assignments?

**Comment by desagas** [comment]
Are you using it in conjunction with a server? And, how exactly can you tell it is messing up the json file if it is the exact same file each time, as you so mentioned? That sounds like it is working perfectly, but somehow the mods are conflicting, with other biome mods, not with this mod. Did you read the “how to install” part of the description on curse, or the mini wiki? It does not fix old worlds, it preserves them, and it doesn’t change biomes from one to the other until after one is loaded, and a duplicate is already found. It does not assign ids unless a prior conflict is caught, like two ids of 214 or something, then it will select the first biome registered, assign that the id, and assign the second biome to another if that becomes available.

So, I assume that may be the issue. You are using multiple mods with biome ids that have been somehow pre assigned within the mod itself, which is bad programming on their behalf. What mods are you using. You literally gave me no information to go on what so ever, like what version, what biome mods you are using, server or single, a copy of your json, etc.

Once you read over my wiki, post those things and let me know.

**Reply by EricLWeiss** [reply]
First, THANKS for your prompt response.  Second, and to be clear: this is NOT some malfunction of biomeidfixer, which has a great record of working exceptionally well for the task it addresses.  I didn't mean to imply any problem with the mod itself.

After doing a bit more inquiry, here's the most precise an order of operations I can deduce.  The world in question was developed in single player with an extensive custom modpack, and then uploaded to a server for multiplayer use (along with the modpack).  Upon running it in multiplayer mode from the server the biomes were scrambled.  The world and modpack were then deleted from the server. Importantly, I'm not clear if biomeidfixer was run on the server copy of the modpack/world prior to this deletion.  However, based on your answer this seems likely to me.

At this point I came into the picture, installed biomeidfixer in the modpack as detailed, and started the world up in single player mode.  As expected this produced a json file, but as noted the biomes were extensively reassigned.  Removing both the mod and the json file reverts the world back to its original biomes, which I find puzzling.

Screenshots show the list of mods, but I don't want to waste your time on what may be an easy fix.  My sense from your answer and other reading is that biomeidfixer must have been run on the server version of the world.  I assume this results in local modification on the client side that instructs biomeidfixer to reconstitute a json file with a consistent set of incorrect assignments.  Is there a way to reverse this client-side modification such that biomeidfixer will produce a json file that reflects the set of biomes present in the single player world?  Again, happy to provide all information if there isn't an easy top-level fix for this.

Again, thanks for your quick response and for producing a really useful mod that addresses an aggravating problem.
<img width="453" alt="Screen Shot 2023-10-04 at 9 29 42 AM" src="https://github.com/desagas/Biome-Id-Fixer/assets/17819792/acaa73de-cdce-4f55-b60f-1fd0629a401f">
<img width="441" alt="Screen Shot 2023-10-04 at 9 29 17 AM" src="https://github.com/desagas/Biome-Id-Fixer/assets/17819792/3249a368-aa60-4bd0-9231-0044dacec2be">

**Comment by desagas** [comment]
So, the solution is to replace the json created after the server is ram for the first time, yes, replace that with then single player json that was created when you made the SP first run. Once the json is there, it shouldn’t change it. As both biome ids and world data are Minecraft saved, not saved by this mod.

Let me know if you tried that, and how it worked. Also, I didn’t think you were blaming the mod or any mod, I just want to know more so we can fix it.

Denver Leman

On Oct 4, 2023, at 8:54 AM, SilenceHand ***@***.***> wrote:

﻿

First, THANKS for your prompt response. Second, and to be clear: this is NOT some malfunction of biomeidfixer, which has a great record of working exceptionally well for the task it addresses. I didn't mean to imply any problem with the mod itself.

After doing a bit more inquiry, here's the most precise an order of operations I can deduce. The world in question was developed in single player with an extensive custom modpack, and then uploaded to a server for multiplayer use (along with the modpack). Upon running it in multiplayer mode from the server the biomes were scrambled. The world and modpack were then deleted from the server. Importantly, I'm not clear if biomeidfixer was run on the server copy of the modpack/world prior to this deletion. However, based on your answer this seems likely to me.

At this point I came into the picture, installed biomeidfixer in the modpack as detailed, and started the world up in single player mode. As expected this produced a json file, but as noted the biomes were extensively reassigned. Removing both the mod and the json file reverts the world back to its original biomes, which I find puzzling.

Screenshots show the list of mods, but I don't want to waste your time on what may be an easy fix. My sense from your answer and other reading is that biomeidfixer must have been run on the server version of the world. I assume this results in local modification on the client side that instructs biomeidfixer to reconstitute a json file with a consistent set of incorrect assignments. Is there a way to reverse this client-side modification such that biomeidfixer will produce a json file that reflects the set of biomes present in the single player world? Again, happy to provide all information if there isn't an easy top-level fix for this.

Again, thanks for your quick response and for producing a really useful mod that addresses an aggravating problem.
[Screen Shot 2023-10-04 at 9 29 42 AM]<https://user-images.githubusercontent.com/17819792/272611156-acaa73de-cdce-4f55-b60f-1fd0629a401f.png>
[Screen Shot 2023-10-04 at 9 29 17 AM]<https://user-images.githubusercontent.com/17819792/272611171-3249a368-aa60-4bd0-9231-0044dacec2be.png>

**Comment by desagas** [comment]
In short, rerun the exact same list of mods on SP, again, with biomeidfixer installed. Make very sure they are all the same versions of the mods as well, and then Copy the created SP json to the server. And, boom, that should fix your issues. You do not need a json on your client if you are playing on a server, only the server needs it.

**Comment by desagas** [comment]
And I really wish BoP actually would have actually fixed their mod to somehow store biome short names with ids somewhere to compare to prevent this from happening with the limited vanilla code, but, meh, it made me happy and gave me something to do to delve into the horribly obfuscated Minecraft code, line by line, adding “hello world” prints at tens of thousands of places until finally I could find where biomes were registered, because the community had not added their own deobfuscated versions yet, where top mod makes told me not to touch biomes, ever bahahah, so no one would help. Ugh. Lol. Hahah.
